elements/lisk-transactions/src/14_unlock_transaction.ts

```diff
@@ -0,0 +1,307 @@
+		const addressArray = this.asset.unlockingObjects.map(unlock => ({
+			address: unlock.delegateAddress,
+		}));
+		const filterArray = [
+			{
+				address: this.senderId,
+			},
+			...addressArray,
+		];
+
+		await store.account.cache(filterArray);
+	}
+
+	protected validateAsset(): ReadonlyArray<TransactionError> {
+		const asset = this.assetToJSON();
+		const schemaErrors = validator.validate(voteAssetFormatSchema, asset);
+		const errors = convertToAssetError(
+			this.id,
+			schemaErrors,
+		) as TransactionError[];
+
+		for (const unlock of this.asset.unlockingObjects) {
+			if (unlock.amount <= BigInt(0)) {
+				errors.push(
+					new TransactionError(
+						'Amount cannot be less than zero',
+						this.id,
+						'.asset.votes.amount',
+						unlock.amount.toString(),
+					),
+				);
+			}
+
+			if (unlock.amount % TEN_UNIT !== BigInt(0)) {
+				errors.push(
+					new TransactionError(
+						'Amount should be multiple of 10 * 10^8',
+						this.id,
+						'.asset.votes.amount',
+						unlock.amount.toString(),
+					),
+				);
+			}
+		}
+
+		return errors;
+	}
+
+	protected async applyAsset(
+		store: StateStore,
+	): Promise<ReadonlyArray<TransactionError>> {
+		const errors = [];
+
+		for (const unlock of this.asset.unlockingObjects) {
+			const sender = await store.account.get(this.senderId);
+			const delegate = await store.account.get(unlock.delegateAddress);
+			if (
+				!hasWaited(sender, delegate, store.chain.lastBlockHeader.height, unlock)
+			) {
+				errors.push(
+					new TransactionError(
+						'Unlocking object has not waited locking period',
+						this.id,
+						'.asset.unlockingObjects.unvoteHeight',
+						unlock.unvoteHeight,
+					),
+				);
+			}
+			if (isPunished(sender, delegate, store.chain.lastBlockHeader.height)) {
+				errors.push(
+					new TransactionError(
+						'Delegate is currently being punished',
+						this.id,
+						'.asset.unlockingObjects.delegateAddress',
+						unlock.unvoteHeight,
+					),
+				);
+			}
+			const unlockIndex = sender.unlocking.findIndex(
+				obj =>
+					obj.amount === unlock.amount &&
+					obj.delegateAddress === unlock.delegateAddress &&
+					obj.unvoteHeight === unlock.unvoteHeight,
+			);
+			if (unlockIndex < 0) {
+				errors.push(
+					new TransactionError(
+						'Corresponding unlocking object not found',
+						this.id,
+						'.asset.unlockingObjects',
+					),
+				);
+				continue;
+			}
+			sender.unlocking.splice(unlockIndex, 1);
+			sender.balance += unlock.amount;
+			store.account.set(sender.address, sender);
+		}
+
+		return errors;
+	}
+
+	protected async undoAsset(
+		store: StateStore,
+	): Promise<ReadonlyArray<TransactionError>> {
+		for (const unlock of this.asset.unlockingObjects) {
+			const sender = await store.account.get(this.senderId);
+
+			sender.balance -= unlock.amount;
+			sender.unlocking.push(unlock);
+			// Resort the unlocking since it's pushed to the end
+			sender.unlocking.sort((a, b) =>
```

Comment:
to account for potentially equal addresses, we should sort by lexicographical order of (address, unvoteHeight, amout)

Reply:
Addressed